packages/client/src/index.ts

```diff
@@ -31,34 +38,68 @@ export const PenumbraSymbol = Symbol.for('penumbra');
  *
  */
 export interface PenumbraInjection {
-  /** Call when creating a channel transport to this provider.  Returns a promise
-   * that may resolve with an active `MessagePort`. */
+  /** Should contain a URI at the provider's origin, serving a manifest
+   * describing this provider. */
+  readonly manifest: string;
+
+  /** Call to acquire a `MessagePort` to this provider, subject to approval. */
   readonly connect: () => Promise<MessagePort>;
 
-  /** Call to gain approval to connect.  Returns a `Promise<void>` that may
-   * reject with an enumerated failure reason. */
+  /** Call to gain approval.  May reject with a `PenumbraProviderRequestError`
+   * containing an enumerated `PenumbraRequestFailure` cause. */
   readonly request: () => Promise<void>;
 
-  /** Call to indicate the provider should revoke approval of this origin. */
+  /** Call to indicate the provider should discard approval of this origin. */
   readonly disconnect: () => Promise<void>;
 
   /** Should synchronously return the present connection state.
-   *
    * - `true` indicates active connection.
    * - `false` indicates connection is closed or rejected.
-   * - `undefined` indicates connection may be attempted.
+   * - `undefined` no attempt has resolved. connection may be attempted.
    */
   readonly isConnected: () => boolean | undefined;
 
-  /** Should contain a URI at the provider's origin, which returns a chrome
-   * extension manifest v3 describing this provider. */
-  readonly manifest: string;
+  /** Synchronously return present injection state. */
+  readonly state: () => PenumbraInjectionState;
+
+  /** Emits `PenubraInjectionStateEvent` when state changes. Listen for
+   * `'penumbrastate'` events, and check the `detail` field for a
+   * `PenumbraInjectionState` value. */
+  readonly addEventListener: EventTarget['addEventListener'];
+  readonly removeEventListener: EventTarget['removeEventListener'];
+}
+
+export class PenumbraInjectionStateEvent extends CustomEvent<PenumbraInjectionState> {
+  constructor(override readonly detail: PenumbraInjectionState) {
+    super('penumbrastate', { detail });
+  }
+}
+
+export enum PenumbraInjectionState {
+  /* error is present */
+  'Failed' = 'Failed',
+
+  /* no action has been taken */
+  'Present' = 'Present',
+
+  /* approval request pending */
+  'RequestPending' = 'RequestPending',
+  /* request for approval satisfied */
+  'Requested' = 'Requested',
+
+  /* connection attempt pending */
+  'ConnectPending' = 'ConnectPending',
+  /* connection successful and active */
+  'Connected' = 'Connected',
+
+  /* disconnect was called to release approval */
+  'Disconnected' = 'Disconnected',
```

Comment:
Instead of emitting strings, what if we just emit a snapshot of the state? Aka:

```ts
enum LifecycleState {
  NOT_REQUESTED,
  PENDING,
  SUCCESS,
  ERROR,
}

interface PenumbraInjectionState {
  connect: LifecycleState;
  request: LifecycleState;
  disconnect: LifecycleState;
}
```

Reply:
Is this question about the return types for injected functions? Right now, they return `Promise<void>` or throw error if something happens.

As for emitting strings vs enum numbers, I personally lean towards strings as the more readable option

Reply:
It's not related to the return types, but a way for someone to know in a subscription-like way what the state of all the possible requests are. And given each method has 4 states, in theory this would be 12 different strings possibly emitted. And then, the consumer of this would need to have their own state object or useState() variables where they take these 12 different strings and have to update each useState() individually.

Reply:
> As for emitting strings vs enum numbers, I personally lean towards strings as the more readable option

this is a string enum - it does not become numbers, it essentially compiles to strings. returned values and parameter values exist in JS as strings and in typescript it provides some convenient collection testing possibilities like `if (someString in PenumbraInjectionState)`

Reply:
>Instead of emitting strings, what if we just emit a snapshot of the state? Aka:

generally i think it's bad to track and expose the internal state of this injection in an explicit or detailed way. doing this restricts implementation possibilities, and expecting a consumer to inspect and interpret some internal state of the injection which may subtly change version-to-version or differ implementation-to-implementation seems unreliable.

most state a caller is interested in is evident by the return values or rejections of the methods they are using. i think it is reasonable to expect the caller to independently understand what actions they have taken, and to use return values and thrown errors.

any state tracking method implemented here is, from the caller's perspective, mostly just answering "what have i already done?" and "did it work?", both of which a caller knows if they are using return values. for synchronous inspection the original boolean-or-undefined state would have satisfied this situation. 

>And given each method has 4 states, in theory this would be 12 different strings possibly emitted. And then, the consumer of this would need to have their own state object or useState() variables where they take these 12 different strings and have to update each useState() individually.

there are three effective methods (request, connect, disconnect) with four effective call states (uncalled, pending, resolved, rejected) meaning there are 64 possible combinations, most of which are invalid and unreachable. if we expect the caller does not know what actions their own code has taken, more detailed state inquiry is more helpful to them, but representing invalid or impossible states seems not helpful.

if the objective here is to simplify use of this interface, expanding from a simple truthy/falsy state to a matrix of 64 mostly invalid states seems like a step backwards. so in the interest of simplification the enum only identifies valid states, and nonsensical states are represented by `Failed` and rejected promises.

valid transitions are like

```mermaid
stateDiagram-v2
direction LR
    Present
    Present --> ConnectionPending
    ConnectionPending --> Connected
    Connected --> Disconnected
    Present --> RequestPending
    RequestPending --> Requested
    Requested --> ConnectionPending
    Connected --> [*]
    Disconnected --> [*]
    note right of Connected
        isConnected(): true
    end note
    note left of Disconnected
        isConnected(): false
    end note
```

and there is not really room for deviation from this path. and there are already meaningful and specific error types for the reachable failure states.

Reply:
> a way for someone to know in a subscription-like way what the state of all the possible requests are.

this is satisfied by the `addEventListener` method now exposed

Reply:
@turbocrime this flow diagram should definitely go to the client package readme or docs. It's great 👍